### ISSUE TITLE
Analyze and improve the performance of Import/export resolution

### ISSUE THREAD

**Issue by Akirathan** [enso-org/enso#9236]
Related to #9235. Benchmark introduced in #9158.

The following benchmark:
```
sbt:runtime-benchmarks> benchOnly org.enso.compiler.benchmarks.module.ImportStandardLibrariesBenchmark
```
measure mostly the performance of import/export resolution, as this screenshot suggests:
![image](https://github.com/enso-org/enso/assets/14013887/7395c155-5648-47d8-841d-7d21148ffe2e)

**Comment by JaroslavTulach** [comment]
While this PR started as a request to speed up a benchmark, I did a real profiling of `./runner` (generated by native image from #9997). While executing `./runner --run test/Base_Tests ^Text` I see more than a second spend in the _import/export_ resolution:

![runImportsAndExportsResolution](https://github.com/enso-org/enso/assets/26887752/3721f906-0919-4422-b3a5-382016c8cec4)

Because this is _native image_, there is no _JVM initialization_ overhead and this is a real cost of our algorithm. _Note: There is a significant cost of GC, but that may be because we generate too much garbage_ somewhere.

How many imports we have? Few thousands? Why should it take to more than a second to process such a small amount of data?

### Analyze

To analyze the data, download [Oracle Developer Studio 12.6](https://docs.oracle.com/en/operating-systems/studio.html) and use its [Analyzer](https://docs.oracle.com/cd/E77782_01/html/E77798/index.html). Download [my experiment data](http://xelfi.cz/enso/flames/runTextTests.zip). Unzip and open with:
```bash
$ PATH=/jdk8/bin:$PATH
$ /OracleDeveloperStudio12.6-linux-x86-bin/developerstudio12.6/bin/analyzer runTextTests.er/
```
flamegraphs view and _find_ (top right corner) shall be your friends when exploring the gathered data.

### Record

Rebuild `./runner` executable with debug information on (otherwise the names methods wouldn't be visible):
```diff
diff --git build.sbt build.sbt
index 0be71a204a..18a5825c87 100644
--- build.sbt
+++ build.sbt
@@ -2432,9 +2432,9 @@ lazy val `engine-runner` = project
             "-Dorg.apache.commons.logging.Log=org.apache.commons.logging.impl.NoOpLog",
             "-H:IncludeResources=.*Main.enso$",
             // useful perf & debug switches:
-            // "-g",
-            // "-H:+SourceLevelDebug",
-            // "-H:-DeleteLocalSymbols",
+            "-g",
+            "-H:+SourceLevelDebug",
+            "-H:-DeleteLocalSymbols",
             // you may need to set smallJdk := None to use following flags:
             // "--trace-class-initialization=org.enso.syntax2.Parser",
             "-Dnic=nic"
```
launch `OracleDeveloperStudio12.6-linux-x86-bin/developerstudio12.6/bin/analyzer` and choose to run an _experiment_. Specify `./runner` as executable and add arguments. Run, save, analyze.

Btw. to speed working with NI up, it may make sense to comment out GraalPython:
```diff
diff --git project/GraalVM.scala project/GraalVM.scala
index 85c6124b76..56b8faea9b 100644
--- project/GraalVM.scala
+++ project/GraalVM.scala
@@ -109,7 +109,7 @@ object GraalVM {
     if (isEspressoMode()) {
       espressoPkgs
     } else {
-      jsPkgs ++ pythonPkgs
+      jsPkgs
     }
 
   private val allowedJavaVendors = Seq(
```

**Comment by JaroslavTulach** [comment]
Do you know why `sbt buildEngineDistribution` takes so long, especially:
```
Generating index for built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/lib/Standard/Table
```
This is the result of sampling:


![Image](https://github.com/enso-org/enso/assets/26887752/e3262c6a-1de4-4946-9222-ea0130775872)

**14s** in `optimize`!?

**Comment by Akirathan** [comment]
Closed by #10369
